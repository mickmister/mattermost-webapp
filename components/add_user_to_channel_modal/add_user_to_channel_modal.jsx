// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import PropTypes from 'prop-types';
import React from 'react';
import {Modal} from 'react-bootstrap';
import {FormattedMessage} from 'react-intl';

import * as UserUtils from 'mattermost-redux/utils/user_utils';

import SearchChannelWithPermissionsProvider from 'components/suggestion/search_channel_with_permissions_provider.jsx';
import SuggestionBox from 'components/suggestion/suggestion_box.jsx';
import SuggestionList from 'components/suggestion/suggestion_list.jsx';

import * as Utils from 'utils/utils.jsx';

export default class AddUserToChannelModal extends React.Component {
    static propTypes = {

        /**
         * Function that's called when modal is closed
         */
        onHide: PropTypes.func.isRequired,

        /**
         * The user that is being added to a channel
         */
        user: PropTypes.object.isRequired,

        /**
         * Object used to determine if the user
         * is a member of a given channel
         */
        channelMembers: PropTypes.object.isRequired,

        actions: PropTypes.shape({

            /**
             * Function to add the user to a channel
             */
            addChannelMember: PropTypes.func.isRequired,

            /**
             * Function to fetch the user's channel membership
             */
            getChannelMember: PropTypes.func.isRequired,
        }).isRequired,
    }

    constructor(props) {
        super(props);

        this.state = {
            show: true,
            saving: false,
            loadingChannels: false,
            checkingForMembership: false,
            text: '',
            selectedChannel: null,
            inviteError: '',
        };
        this.suggestionProviders = [new SearchChannelWithPermissionsProvider()];
        this.enableChannelProvider();
    }

    enableChannelProvider = () => {
        this.suggestionProviders[0].disableDispatches = false;
    }

    focusTextbox = () => {
        if (this.channelSearchBox == null) {
            return;
        }

        const textbox = this.channelSearchBox.getTextbox();
        if (document.activeElement !== textbox) {
            textbox.focus();
            Utils.placeCaretAtEnd(textbox);
        }
    }

    onChange = (e) => {
        this.setState({text: e.target.value, selectedChannel: null});
    }

    onHide = () => {
        this.setState({show: false});
    }

    setSearchBoxRef = (input) => {
        this.channelSearchBox = input;
        this.focusTextbox();
    }

    handleInviteError = (error) => {
        if (error) {
            this.setState({inviteError: error.message, saving: false});
            return;
        }
        this.onHide();
    }

    didSelectChannel = (selection) => {
        const channel = selection.channel;
        const userId = this.props.user.id;

        this.setState({
            text: channel.display_name,
            selectedChannel: channel,
            checkingForMembership: true,
            inviteError: '',
        });

        this.props.actions.getChannelMember(channel.id, userId).then(() => {
            this.setState({checkingForMembership: false});
        });
    }

    handleSubmit = () => {
        const channel = this.state.selectedChannel;
        const user = this.props.user;

        if (!channel) {
            return;
        }

        this.setState({saving: true});

        this.props.actions.addChannelMember(channel.id, user.id).then(({error}) => {
            this.handleInviteError(error);
        });
    }

    isUserMemberOfChannel = (channel) => {
        const user = this.props.user;
        const memberships = this.props.channelMembers;

        if (!channel) {
            return false;
        }

        if (!memberships[channel.id]) {
            return false;
        }

        return Boolean(memberships[channel.id][user.id]);
    }

    render() {
        const user = this.props.user;
        const channel = this.state.selectedChannel;
        const targetUserIsMemberOfSelectedChannel = this.isUserMemberOfChannel(channel);

        let name = UserUtils.getFullName(user);
        if (!name) {
            name = `@${user.username}`;
        }

        let inviteError;
        if (!this.state.saving) {
            if (this.state.inviteError) {
                inviteError = (<label className='has-error control-label'>{this.state.inviteError}</label>);
            } else if (targetUserIsMemberOfSelectedChannel) {
                inviteError = (<label className='has-error control-label'>{`${name} is already a member of that channel`}</label>);
            }
        }

        const help = (
            <FormattedMessage
                id='add_user_to_channel_modal.help'
                defaultMessage='Type to find a channel. Use ↑↓ to browse, ↵ to select, ESC to dismiss.'
            />
        );

        const content = (
            <SuggestionBox
                ref={this.setSearchBoxRef}
                className='form-control focused'
                onChange={this.onChange}
                value={this.state.text}
                onKeyDown={this.handleKeyDown}
                onItemSelected={this.didSelectChannel}
                listComponent={SuggestionList}
                maxLength='64'
                providers={this.suggestionProviders}
                listStyle='bottom'
                completeOnTab={false}
                renderDividers={false}
                delayInputUpdate={true}
                openWhenEmpty={true}
            />
        );

        const shouldDisableAddButton = targetUserIsMemberOfSelectedChannel ||
            this.state.checkingForMembership ||
            Boolean(!this.state.selectedChannel) ||
            this.state.saving;

        return (
            <Modal
                dialogClassName='modal--overflow'
                show={this.state.show}
                onHide={this.onHide}
                onExited={this.props.onHide}
                ref='modal'
                enforceFocus={true}
            >
                <Modal.Header closeButton={true}>
                    <Modal.Title>
                        <FormattedMessage
                            id='add_user_to_channel.title'
                            defaultMessage='Add {name} to a channel.'
                            values={{
                                name,
                            }}
                        />
                    </Modal.Title>
                </Modal.Header>
                <Modal.Body>
                    <div className='modal__hint'>
                        {help}
                    </div>
                    {content}
                    <div className='form-group has-error'>
                        <br/>
                        {inviteError}
                    </div>
                </Modal.Body>
                <Modal.Footer>
                    <button
                        type='button'
                        className='btn btn-link'
                        onClick={this.onHide}
                    >
                        <FormattedMessage
                            id='add_user_to_channel_modal.cancel'
                            defaultMessage='Cancel'
                        />
                    </button>
                    <button
                        type='button'
                        className='btn btn-primary'
                        onClick={this.handleSubmit}
                        disabled={shouldDisableAddButton}
                    >
                        <FormattedMessage
                            id='add_user_to_channel_modal.add'
                            defaultMessage='Add'
                        />
                    </button>
                </Modal.Footer>
            </Modal>
        );
    }
}
