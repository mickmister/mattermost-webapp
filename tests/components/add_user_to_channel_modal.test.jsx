// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import React from 'react';
import {shallow} from 'enzyme';

import {Child as AddUserToChannelModal} from 'components/add_user_to_channel_modal';

describe('components/AddUserToChannelModal', () => {
    const baseProps = {
        channelMembers: {},
        user: {id: 'someUserId'},
        onHide: jest.fn(),
        actions: {
            addChannelMember: jest.fn(() => {
                return new Promise((resolve) => {
                    process.nextTick(() => resolve());
                });
            }),
            getChannelMember: jest.fn(() => {
                return new Promise((resolve) => {
                    process.nextTick(() => resolve());
                });
            }),
        },
    };

    it('should match snapshot', () => {
        const wrapper = shallow(
            <AddUserToChannelModal {...baseProps}/>
        );

        expect(wrapper).toMatchSnapshot();
    });

    it('should match snapshot when a channel is selected', () => {
        const wrapper = shallow(
            <AddUserToChannelModal {...baseProps}/>
        );

        wrapper.setState({inviteError: 'some error'});
        expect(wrapper).toMatchSnapshot();

        wrapper.setState({selectedChannelId: 'someChannelId'});
        expect(wrapper).toMatchSnapshot();
    });

    it('should match snapshot when an error message is captured', () => {
        const wrapper = shallow(
            <AddUserToChannelModal {...baseProps}/>
        );

        wrapper.setState({inviteError: 'some error'});
        expect(wrapper).toMatchSnapshot();
    });

    it('should match snapshot when membership is being checked', () => {
        const wrapper = shallow(
            <AddUserToChannelModal {...baseProps}/>
        );

        wrapper.setState({
            selectedChannelId: 'someChannelId',
            checkingForMembership: true,
        });
        expect(wrapper).toMatchSnapshot();
    });

    it('should match snapshot if user is a member of the selected channel', () => {
        const props = {...baseProps,
            channelMembers: {
                someChannelId: {
                    someUserId: {},
                },
            },
        };

        const wrapper = shallow(
            <AddUserToChannelModal {...props}/>
        );

        wrapper.setState({selectedChannelId: 'someChannelId'});
        expect(wrapper).toMatchSnapshot();
    });

    it('should match snapshot when saving', () => {
        const wrapper = shallow(
            <AddUserToChannelModal {...baseProps}/>
        );

        wrapper.setState({
            selectedChannelId: 'someChannelId',
            saving: true,
        });
        expect(wrapper).toMatchSnapshot();
    });

    describe('didSelectChannel', () => {
        it('should fetch the selected user\'s membership for the selected channel', () => {
            const props = {...baseProps};

            const wrapper = shallow(
                <AddUserToChannelModal {...props}/>
            );

            const selection = {channel: {id: 'someChannelId', display_name: 'channelName'}};
            wrapper.instance().didSelectChannel(selection);
            expect(props.actions.getChannelMember).toBeCalledWith('someChannelId', 'someUserId');
        });

        it('should match state on selection', async () => {
            const promise = Promise.resolve({});
            const props = {
                ...baseProps,
                actions: {
                    ...baseProps.actions,
                    getChannelMember: jest.fn(() => {
                        return promise;
                    }),
                },
            };

            const wrapper = shallow(
                <AddUserToChannelModal {...props}/>
            );

            expect(wrapper.state().text).toEqual('');
            expect(wrapper.state().checkingForMembership).toEqual(false);
            expect(wrapper.state().selectedChannelId).toEqual(null);
            expect(wrapper.state().inviteError).toEqual('');

            const selection = {channel: {id: 'someChannelId', display_name: 'channelName'}};
            wrapper.setState({inviteError: 'some pre-existing error'});

            wrapper.instance().didSelectChannel(selection);
            expect(wrapper.state().text).toEqual('channelName');
            expect(wrapper.state().checkingForMembership).toEqual(true);
            expect(wrapper.state().selectedChannelId).toEqual('someChannelId');
            expect(wrapper.state().inviteError).toEqual('');

            await promise;
            expect(wrapper.state().checkingForMembership).toEqual(false);
        });
    });

    describe('handleSubmit', () => {
        it('should do nothing if no channel is selected', () => {
            const props = {...baseProps};

            const wrapper = shallow(
                <AddUserToChannelModal {...props}/>
            );

            wrapper.instance().handleSubmit();
            expect(wrapper.state().saving).toBe(false);
            expect(props.actions.addChannelMember).not.toBeCalled();
        });

        it('should do nothing if user is a member of the selected channel', () => {
            const props = {...baseProps,
                channelMembers: {
                    someChannelId: {
                        someUserId: {},
                    },
                },
            };

            const wrapper = shallow(
                <AddUserToChannelModal {...props}/>
            );

            wrapper.setState({selectedChannelId: 'someChannelId'});
            wrapper.instance().handleSubmit();
            expect(wrapper.state().saving).toBe(false);
            expect(props.actions.addChannelMember).not.toBeCalled();
        });

        it('should submit if user is not a member of the selected channel', () => {
            const props = {...baseProps,
                channelMembers: {
                    someChannelId: {},
                },
            };

            const wrapper = shallow(
                <AddUserToChannelModal {...props}/>
            );

            wrapper.setState({selectedChannelId: 'someChannelId'});
            wrapper.instance().handleSubmit();
            expect(wrapper.state().saving).toBe(true);
            expect(props.actions.addChannelMember).toBeCalled();
        });

        test('should match state when save is successful', async () => {
            const onHide = jest.fn();
            const promise = Promise.resolve({});
            const props = {
                ...baseProps,
                onHide,
                actions: {
                    ...baseProps.actions,
                    addChannelMember: () => promise,
                },
            };

            const wrapper = shallow(
                <AddUserToChannelModal {...props}/>
            );

            expect(wrapper.state().show).toBe(true);
            expect(wrapper.state().saving).toBe(false);
            wrapper.setState({selectedChannelId: 'someChannelId'});

            wrapper.instance().handleSubmit();
            expect(wrapper.state().show).toBe(true);
            expect(wrapper.state().saving).toBe(true);

            await promise;
            expect(wrapper.state().inviteError).toEqual('');
            expect(wrapper.state().show).toBe(false);
            expect(onHide).toHaveBeenCalled();
        });

        test('should match state when save fails', async () => {
            const onHide = jest.fn();
            const promise = Promise.resolve({error: new Error('some error')});
            const props = {
                ...baseProps,
                onHide,
                actions: {
                    ...baseProps.actions,
                    addChannelMember: () => promise,
                },
            };

            const wrapper = shallow(
                <AddUserToChannelModal {...props}/>
            );

            expect(wrapper.state().show).toBe(true);
            wrapper.setState({selectedChannelId: 'someChannelId'});

            wrapper.instance().handleSubmit();

            await promise;
            expect(wrapper.state().inviteError).toEqual('some error');
            expect(wrapper.state().show).toBe(true);
            expect(onHide).not.toHaveBeenCalled();
        });
    });
});
