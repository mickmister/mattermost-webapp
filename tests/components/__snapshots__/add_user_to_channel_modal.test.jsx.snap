// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/AddUserToChannelModal should match snapshot 1`] = `
<Modal
  animation={true}
  autoFocus={true}
  backdrop={true}
  bsClass="modal"
  dialogClassName="modal--overflow"
  dialogComponentClass={[Function]}
  enforceFocus={true}
  keyboard={true}
  manager={
    ModalManager {
      "add": [Function],
      "containers": Array [],
      "data": Array [],
      "handleContainerOverflow": true,
      "hideSiblingNodes": true,
      "isTopModal": [Function],
      "modals": Array [],
      "remove": [Function],
    }
  }
  onExited={[MockFunction]}
  onHide={[Function]}
  renderBackdrop={[Function]}
  restoreFocus={true}
  show={true}
>
  <ModalHeader
    bsClass="modal-header"
    closeButton={true}
    closeLabel="Close"
  >
    <ModalTitle
      bsClass="modal-title"
      componentClass="h4"
    >
      <FormattedMessage
        defaultMessage="Add {name} to a channel"
        id="add_user_to_channel_modal.title"
        values={
          Object {
            "name": "@undefined",
          }
        }
      />
    </ModalTitle>
  </ModalHeader>
  <ModalBody
    bsClass="modal-body"
    componentClass="div"
  >
    <div
      className="modal__hint"
    >
      <FormattedMessage
        defaultMessage="Type to find a channel. Use ↑↓ to browse, ↵ to select, ESC to dismiss."
        id="add_user_to_channel_modal.help"
        values={Object {}}
      />
    </div>
    <SuggestionBox
      className="form-control focused"
      completeOnTab={false}
      containerClass=""
      delayInputUpdate={true}
      isRHS={false}
      listComponent={[Function]}
      listStyle="bottom"
      maxLength="64"
      onChange={[Function]}
      onItemSelected={[Function]}
      openOnFocus={false}
      openWhenEmpty={true}
      providers={
        Array [
          SearchChannelWithPermissionsProvider {
            "appDispatch": [Function],
            "disableDispatches": false,
            "dispatch": [Function],
            "getState": [Function],
            "latestComplete": true,
            "latestPrefix": "",
            "requestStarted": false,
          },
        ]
      }
      renderDividers={false}
      renderNoResults={false}
      replaceAllInputOnSelect={false}
      requiredCharacters={1}
      value=""
    />
    <div>
      <br />
    </div>
  </ModalBody>
  <ModalFooter
    bsClass="modal-footer"
    componentClass="div"
  >
    <button
      className="btn btn-link"
      onClick={[Function]}
      type="button"
    >
      <FormattedMessage
        defaultMessage="Cancel"
        id="add_user_to_channel_modal.cancel"
        values={Object {}}
      />
    </button>
    <button
      className="btn btn-primary"
      disabled={true}
      onClick={[Function]}
      type="button"
    >
      <FormattedMessage
        defaultMessage="Add"
        id="add_user_to_channel_modal.add"
        values={Object {}}
      />
    </button>
  </ModalFooter>
</Modal>
`;

exports[`components/AddUserToChannelModal should match snapshot if user is a member of the selected channel 1`] = `
<Modal
  animation={true}
  autoFocus={true}
  backdrop={true}
  bsClass="modal"
  dialogClassName="modal--overflow"
  dialogComponentClass={[Function]}
  enforceFocus={true}
  keyboard={true}
  manager={
    ModalManager {
      "add": [Function],
      "containers": Array [],
      "data": Array [],
      "handleContainerOverflow": true,
      "hideSiblingNodes": true,
      "isTopModal": [Function],
      "modals": Array [],
      "remove": [Function],
    }
  }
  onExited={[MockFunction]}
  onHide={[Function]}
  renderBackdrop={[Function]}
  restoreFocus={true}
  show={true}
>
  <ModalHeader
    bsClass="modal-header"
    closeButton={true}
    closeLabel="Close"
  >
    <ModalTitle
      bsClass="modal-title"
      componentClass="h4"
    >
      <FormattedMessage
        defaultMessage="Add {name} to a channel"
        id="add_user_to_channel_modal.title"
        values={
          Object {
            "name": "@undefined",
          }
        }
      />
    </ModalTitle>
  </ModalHeader>
  <ModalBody
    bsClass="modal-body"
    componentClass="div"
  >
    <div
      className="modal__hint"
    >
      <FormattedMessage
        defaultMessage="Type to find a channel. Use ↑↓ to browse, ↵ to select, ESC to dismiss."
        id="add_user_to_channel_modal.help"
        values={Object {}}
      />
    </div>
    <SuggestionBox
      className="form-control focused"
      completeOnTab={false}
      containerClass=""
      delayInputUpdate={true}
      isRHS={false}
      listComponent={[Function]}
      listStyle="bottom"
      maxLength="64"
      onChange={[Function]}
      onItemSelected={[Function]}
      openOnFocus={false}
      openWhenEmpty={true}
      providers={
        Array [
          SearchChannelWithPermissionsProvider {
            "appDispatch": [Function],
            "disableDispatches": false,
            "dispatch": [Function],
            "getState": [Function],
            "latestComplete": true,
            "latestPrefix": "",
            "requestStarted": false,
          },
        ]
      }
      renderDividers={false}
      renderNoResults={false}
      replaceAllInputOnSelect={false}
      requiredCharacters={1}
      value=""
    />
    <div>
      <label
        className="modal__error has-error control-label"
      >
        <FormattedMessage
          defaultMessage="{name} is already a member of that channel"
          id="add_user_to_channel_modal.membershipExistsError"
          values={
            Object {
              "name": "@undefined",
            }
          }
        />
      </label>
      <br />
    </div>
  </ModalBody>
  <ModalFooter
    bsClass="modal-footer"
    componentClass="div"
  >
    <button
      className="btn btn-link"
      onClick={[Function]}
      type="button"
    >
      <FormattedMessage
        defaultMessage="Cancel"
        id="add_user_to_channel_modal.cancel"
        values={Object {}}
      />
    </button>
    <button
      className="btn btn-primary"
      disabled={true}
      onClick={[Function]}
      type="button"
    >
      <FormattedMessage
        defaultMessage="Add"
        id="add_user_to_channel_modal.add"
        values={Object {}}
      />
    </button>
  </ModalFooter>
</Modal>
`;

exports[`components/AddUserToChannelModal should match snapshot when a channel is selected 1`] = `
<Modal
  animation={true}
  autoFocus={true}
  backdrop={true}
  bsClass="modal"
  dialogClassName="modal--overflow"
  dialogComponentClass={[Function]}
  enforceFocus={true}
  keyboard={true}
  manager={
    ModalManager {
      "add": [Function],
      "containers": Array [],
      "data": Array [],
      "handleContainerOverflow": true,
      "hideSiblingNodes": true,
      "isTopModal": [Function],
      "modals": Array [],
      "remove": [Function],
    }
  }
  onExited={[MockFunction]}
  onHide={[Function]}
  renderBackdrop={[Function]}
  restoreFocus={true}
  show={true}
>
  <ModalHeader
    bsClass="modal-header"
    closeButton={true}
    closeLabel="Close"
  >
    <ModalTitle
      bsClass="modal-title"
      componentClass="h4"
    >
      <FormattedMessage
        defaultMessage="Add {name} to a channel"
        id="add_user_to_channel_modal.title"
        values={
          Object {
            "name": "@undefined",
          }
        }
      />
    </ModalTitle>
  </ModalHeader>
  <ModalBody
    bsClass="modal-body"
    componentClass="div"
  >
    <div
      className="modal__hint"
    >
      <FormattedMessage
        defaultMessage="Type to find a channel. Use ↑↓ to browse, ↵ to select, ESC to dismiss."
        id="add_user_to_channel_modal.help"
        values={Object {}}
      />
    </div>
    <SuggestionBox
      className="form-control focused"
      completeOnTab={false}
      containerClass=""
      delayInputUpdate={true}
      isRHS={false}
      listComponent={[Function]}
      listStyle="bottom"
      maxLength="64"
      onChange={[Function]}
      onItemSelected={[Function]}
      openOnFocus={false}
      openWhenEmpty={true}
      providers={
        Array [
          SearchChannelWithPermissionsProvider {
            "appDispatch": [Function],
            "disableDispatches": false,
            "dispatch": [Function],
            "getState": [Function],
            "latestComplete": true,
            "latestPrefix": "",
            "requestStarted": false,
          },
        ]
      }
      renderDividers={false}
      renderNoResults={false}
      replaceAllInputOnSelect={false}
      requiredCharacters={1}
      value=""
    />
    <div>
      <label
        className="modal__error has-error control-label"
      >
        some error
      </label>
      <br />
    </div>
  </ModalBody>
  <ModalFooter
    bsClass="modal-footer"
    componentClass="div"
  >
    <button
      className="btn btn-link"
      onClick={[Function]}
      type="button"
    >
      <FormattedMessage
        defaultMessage="Cancel"
        id="add_user_to_channel_modal.cancel"
        values={Object {}}
      />
    </button>
    <button
      className="btn btn-primary"
      disabled={true}
      onClick={[Function]}
      type="button"
    >
      <FormattedMessage
        defaultMessage="Add"
        id="add_user_to_channel_modal.add"
        values={Object {}}
      />
    </button>
  </ModalFooter>
</Modal>
`;

exports[`components/AddUserToChannelModal should match snapshot when a channel is selected 2`] = `
<Modal
  animation={true}
  autoFocus={true}
  backdrop={true}
  bsClass="modal"
  dialogClassName="modal--overflow"
  dialogComponentClass={[Function]}
  enforceFocus={true}
  keyboard={true}
  manager={
    ModalManager {
      "add": [Function],
      "containers": Array [],
      "data": Array [],
      "handleContainerOverflow": true,
      "hideSiblingNodes": true,
      "isTopModal": [Function],
      "modals": Array [],
      "remove": [Function],
    }
  }
  onExited={[MockFunction]}
  onHide={[Function]}
  renderBackdrop={[Function]}
  restoreFocus={true}
  show={true}
>
  <ModalHeader
    bsClass="modal-header"
    closeButton={true}
    closeLabel="Close"
  >
    <ModalTitle
      bsClass="modal-title"
      componentClass="h4"
    >
      <FormattedMessage
        defaultMessage="Add {name} to a channel"
        id="add_user_to_channel_modal.title"
        values={
          Object {
            "name": "@undefined",
          }
        }
      />
    </ModalTitle>
  </ModalHeader>
  <ModalBody
    bsClass="modal-body"
    componentClass="div"
  >
    <div
      className="modal__hint"
    >
      <FormattedMessage
        defaultMessage="Type to find a channel. Use ↑↓ to browse, ↵ to select, ESC to dismiss."
        id="add_user_to_channel_modal.help"
        values={Object {}}
      />
    </div>
    <SuggestionBox
      className="form-control focused"
      completeOnTab={false}
      containerClass=""
      delayInputUpdate={true}
      isRHS={false}
      listComponent={[Function]}
      listStyle="bottom"
      maxLength="64"
      onChange={[Function]}
      onItemSelected={[Function]}
      openOnFocus={false}
      openWhenEmpty={true}
      providers={
        Array [
          SearchChannelWithPermissionsProvider {
            "appDispatch": [Function],
            "disableDispatches": false,
            "dispatch": [Function],
            "getState": [Function],
            "latestComplete": true,
            "latestPrefix": "",
            "requestStarted": false,
          },
        ]
      }
      renderDividers={false}
      renderNoResults={false}
      replaceAllInputOnSelect={false}
      requiredCharacters={1}
      value=""
    />
    <div>
      <label
        className="modal__error has-error control-label"
      >
        some error
      </label>
      <br />
    </div>
  </ModalBody>
  <ModalFooter
    bsClass="modal-footer"
    componentClass="div"
  >
    <button
      className="btn btn-link"
      onClick={[Function]}
      type="button"
    >
      <FormattedMessage
        defaultMessage="Cancel"
        id="add_user_to_channel_modal.cancel"
        values={Object {}}
      />
    </button>
    <button
      className="btn btn-primary"
      disabled={false}
      onClick={[Function]}
      type="button"
    >
      <FormattedMessage
        defaultMessage="Add"
        id="add_user_to_channel_modal.add"
        values={Object {}}
      />
    </button>
  </ModalFooter>
</Modal>
`;

exports[`components/AddUserToChannelModal should match snapshot when an error message is captured 1`] = `
<Modal
  animation={true}
  autoFocus={true}
  backdrop={true}
  bsClass="modal"
  dialogClassName="modal--overflow"
  dialogComponentClass={[Function]}
  enforceFocus={true}
  keyboard={true}
  manager={
    ModalManager {
      "add": [Function],
      "containers": Array [],
      "data": Array [],
      "handleContainerOverflow": true,
      "hideSiblingNodes": true,
      "isTopModal": [Function],
      "modals": Array [],
      "remove": [Function],
    }
  }
  onExited={[MockFunction]}
  onHide={[Function]}
  renderBackdrop={[Function]}
  restoreFocus={true}
  show={true}
>
  <ModalHeader
    bsClass="modal-header"
    closeButton={true}
    closeLabel="Close"
  >
    <ModalTitle
      bsClass="modal-title"
      componentClass="h4"
    >
      <FormattedMessage
        defaultMessage="Add {name} to a channel"
        id="add_user_to_channel_modal.title"
        values={
          Object {
            "name": "@undefined",
          }
        }
      />
    </ModalTitle>
  </ModalHeader>
  <ModalBody
    bsClass="modal-body"
    componentClass="div"
  >
    <div
      className="modal__hint"
    >
      <FormattedMessage
        defaultMessage="Type to find a channel. Use ↑↓ to browse, ↵ to select, ESC to dismiss."
        id="add_user_to_channel_modal.help"
        values={Object {}}
      />
    </div>
    <SuggestionBox
      className="form-control focused"
      completeOnTab={false}
      containerClass=""
      delayInputUpdate={true}
      isRHS={false}
      listComponent={[Function]}
      listStyle="bottom"
      maxLength="64"
      onChange={[Function]}
      onItemSelected={[Function]}
      openOnFocus={false}
      openWhenEmpty={true}
      providers={
        Array [
          SearchChannelWithPermissionsProvider {
            "appDispatch": [Function],
            "disableDispatches": false,
            "dispatch": [Function],
            "getState": [Function],
            "latestComplete": true,
            "latestPrefix": "",
            "requestStarted": false,
          },
        ]
      }
      renderDividers={false}
      renderNoResults={false}
      replaceAllInputOnSelect={false}
      requiredCharacters={1}
      value=""
    />
    <div>
      <label
        className="modal__error has-error control-label"
      >
        some error
      </label>
      <br />
    </div>
  </ModalBody>
  <ModalFooter
    bsClass="modal-footer"
    componentClass="div"
  >
    <button
      className="btn btn-link"
      onClick={[Function]}
      type="button"
    >
      <FormattedMessage
        defaultMessage="Cancel"
        id="add_user_to_channel_modal.cancel"
        values={Object {}}
      />
    </button>
    <button
      className="btn btn-primary"
      disabled={true}
      onClick={[Function]}
      type="button"
    >
      <FormattedMessage
        defaultMessage="Add"
        id="add_user_to_channel_modal.add"
        values={Object {}}
      />
    </button>
  </ModalFooter>
</Modal>
`;

exports[`components/AddUserToChannelModal should match snapshot when membership is being checked 1`] = `
<Modal
  animation={true}
  autoFocus={true}
  backdrop={true}
  bsClass="modal"
  dialogClassName="modal--overflow"
  dialogComponentClass={[Function]}
  enforceFocus={true}
  keyboard={true}
  manager={
    ModalManager {
      "add": [Function],
      "containers": Array [],
      "data": Array [],
      "handleContainerOverflow": true,
      "hideSiblingNodes": true,
      "isTopModal": [Function],
      "modals": Array [],
      "remove": [Function],
    }
  }
  onExited={[MockFunction]}
  onHide={[Function]}
  renderBackdrop={[Function]}
  restoreFocus={true}
  show={true}
>
  <ModalHeader
    bsClass="modal-header"
    closeButton={true}
    closeLabel="Close"
  >
    <ModalTitle
      bsClass="modal-title"
      componentClass="h4"
    >
      <FormattedMessage
        defaultMessage="Add {name} to a channel"
        id="add_user_to_channel_modal.title"
        values={
          Object {
            "name": "@undefined",
          }
        }
      />
    </ModalTitle>
  </ModalHeader>
  <ModalBody
    bsClass="modal-body"
    componentClass="div"
  >
    <div
      className="modal__hint"
    >
      <FormattedMessage
        defaultMessage="Type to find a channel. Use ↑↓ to browse, ↵ to select, ESC to dismiss."
        id="add_user_to_channel_modal.help"
        values={Object {}}
      />
    </div>
    <SuggestionBox
      className="form-control focused"
      completeOnTab={false}
      containerClass=""
      delayInputUpdate={true}
      isRHS={false}
      listComponent={[Function]}
      listStyle="bottom"
      maxLength="64"
      onChange={[Function]}
      onItemSelected={[Function]}
      openOnFocus={false}
      openWhenEmpty={true}
      providers={
        Array [
          SearchChannelWithPermissionsProvider {
            "appDispatch": [Function],
            "disableDispatches": false,
            "dispatch": [Function],
            "getState": [Function],
            "latestComplete": true,
            "latestPrefix": "",
            "requestStarted": false,
          },
        ]
      }
      renderDividers={false}
      renderNoResults={false}
      replaceAllInputOnSelect={false}
      requiredCharacters={1}
      value=""
    />
    <div>
      <br />
    </div>
  </ModalBody>
  <ModalFooter
    bsClass="modal-footer"
    componentClass="div"
  >
    <button
      className="btn btn-link"
      onClick={[Function]}
      type="button"
    >
      <FormattedMessage
        defaultMessage="Cancel"
        id="add_user_to_channel_modal.cancel"
        values={Object {}}
      />
    </button>
    <button
      className="btn btn-primary"
      disabled={true}
      onClick={[Function]}
      type="button"
    >
      <FormattedMessage
        defaultMessage="Add"
        id="add_user_to_channel_modal.add"
        values={Object {}}
      />
    </button>
  </ModalFooter>
</Modal>
`;

exports[`components/AddUserToChannelModal should match snapshot when saving 1`] = `
<Modal
  animation={true}
  autoFocus={true}
  backdrop={true}
  bsClass="modal"
  dialogClassName="modal--overflow"
  dialogComponentClass={[Function]}
  enforceFocus={true}
  keyboard={true}
  manager={
    ModalManager {
      "add": [Function],
      "containers": Array [],
      "data": Array [],
      "handleContainerOverflow": true,
      "hideSiblingNodes": true,
      "isTopModal": [Function],
      "modals": Array [],
      "remove": [Function],
    }
  }
  onExited={[MockFunction]}
  onHide={[Function]}
  renderBackdrop={[Function]}
  restoreFocus={true}
  show={true}
>
  <ModalHeader
    bsClass="modal-header"
    closeButton={true}
    closeLabel="Close"
  >
    <ModalTitle
      bsClass="modal-title"
      componentClass="h4"
    >
      <FormattedMessage
        defaultMessage="Add {name} to a channel"
        id="add_user_to_channel_modal.title"
        values={
          Object {
            "name": "@undefined",
          }
        }
      />
    </ModalTitle>
  </ModalHeader>
  <ModalBody
    bsClass="modal-body"
    componentClass="div"
  >
    <div
      className="modal__hint"
    >
      <FormattedMessage
        defaultMessage="Type to find a channel. Use ↑↓ to browse, ↵ to select, ESC to dismiss."
        id="add_user_to_channel_modal.help"
        values={Object {}}
      />
    </div>
    <SuggestionBox
      className="form-control focused"
      completeOnTab={false}
      containerClass=""
      delayInputUpdate={true}
      isRHS={false}
      listComponent={[Function]}
      listStyle="bottom"
      maxLength="64"
      onChange={[Function]}
      onItemSelected={[Function]}
      openOnFocus={false}
      openWhenEmpty={true}
      providers={
        Array [
          SearchChannelWithPermissionsProvider {
            "appDispatch": [Function],
            "disableDispatches": false,
            "dispatch": [Function],
            "getState": [Function],
            "latestComplete": true,
            "latestPrefix": "",
            "requestStarted": false,
          },
        ]
      }
      renderDividers={false}
      renderNoResults={false}
      replaceAllInputOnSelect={false}
      requiredCharacters={1}
      value=""
    />
    <div>
      <br />
    </div>
  </ModalBody>
  <ModalFooter
    bsClass="modal-footer"
    componentClass="div"
  >
    <button
      className="btn btn-link"
      onClick={[Function]}
      type="button"
    >
      <FormattedMessage
        defaultMessage="Cancel"
        id="add_user_to_channel_modal.cancel"
        values={Object {}}
      />
    </button>
    <button
      className="btn btn-primary"
      disabled={true}
      onClick={[Function]}
      type="button"
    >
      <FormattedMessage
        defaultMessage="Add"
        id="add_user_to_channel_modal.add"
        values={Object {}}
      />
    </button>
  </ModalFooter>
</Modal>
`;
